### PR TITLE
Adding override parameters to CSVDiff.

### PR DESCRIPTION
**Pull request by acasagran**
Adding override parameters and checks to the CSVDiff tester. Issue #11250.

### CODE REVIEW

**Comment by permcody** [comment]
Tag @brianmoose and @milljm - We need to help Al get some testing going with this PR.

File: python/TestHarness/CSVDiffer.py [hunk 1]
```diff
@@ -10,9 +10,12 @@
 import os, re, math
 
 class CSVDiffer:
-    def __init__(self, test_dir, out_files, abs_zero=1e-11, relative_error=5.5e-6, gold_dir='gold'):
+    def __init__(self, custom_columns, custom_rel_err, custom_abs_zero, test_dir, out_files, abs_zero=1e-11, relative_error=5.5e-6, gold_dir='gold'):
```
Comment (brianmoose) [inline python/TestHarness/CSVDiffer.py]: Can we move these new parameters to the end with a suitable default so existing calls don't break (and nicer calling semantics)? Ie something like:
`def __init__(self, test_dir, out_files, abs_zero=1e-11, relative_error=5.5e-6, gold_dir='gold', custom_columns=[], custom_rel_err=[], custom_abs_zero=[]):`

Maybe put some comments in on how these are to be used...

File: python/TestHarness/CSVDiffer.py [hunk 2]
```diff
@@ -77,6 +90,15 @@ def diff(self):
             if foundError:
                 continue
 
+            # check if custom tolerances used, column name exists in one of
+            # the CSV files
+            if len(self.custom_columns) > 0:
```
Comment (brianmoose) [inline python/TestHarness/CSVDiffer.py]: can just use `if self.custom_columns:`

File: python/TestHarness/CSVDiffer.py [hunk 3]
```diff
@@ -92,7 +114,13 @@ def diff(self):
             rel_tol   = self.rel_tol
             for key in keys1:
                 for val1, val2 in zip( table1[key], table2[key] ):
-                    # adjust to the absolute zero
+                    # if customized tolerances specified use them otherwise
+                    # use the default
+                    if len(self.custom_columns) > 0:
```
Comment (brianmoose) [inline python/TestHarness/CSVDiffer.py]: same

File: python/TestHarness/CSVDiffer.py [hunk 4]
```diff
@@ -114,11 +142,25 @@ def diff(self):
                     if max( abs(val1), abs(val2) ) > 0:
                         rel_diff = abs( ( val1 - val2 ) / max( abs(val1), abs(val2) ) )
 
+                    # if customized tolerances specified use them otherwise
+                    # use the default
+                    if len(self.custom_columns) > 0:
```
Comment (brianmoose) [inline python/TestHarness/CSVDiffer.py]: same

File: python/TestHarness/CSVDiffer.py [hunk 5]
```diff
@@ -114,11 +142,25 @@ def diff(self):
...
+                        except:
+                            rel_tol = self.rel_tol
                     if rel_diff > rel_tol:
                         self.addError(fname, "The values in column \"" + key.strip() + "\" don't match\n\trel diff:  " + str(val1) + " ~ " + str(val2) + " = " + str(rel_diff))
                         # assume all other vals in this column are wrong too, so don't report them
                         break
 
+        # Loop over variable names to check if any are missing from all the
+        # CSV files being compared
+        if len(self.custom_columns) > 0:
```
Comment (brianmoose) [inline python/TestHarness/CSVDiffer.py]: same

**Review by brianmoose — COMMENTED** [review]
There are some existing tests at the bottom of `CSVDiffer` (that never get run). We can move those into some python unittests and then add some more to test this new functionality.

**Review by brianmoose — CHANGES_REQUESTED** [review]
My other pr #11254 got merged so you will probably need a rebase.
You mentioned you had a failing `TestHarness.tests.diff`? Seems to work for me.

File: python/TestHarness/testers/CSVDiff.py [hunk 6]
```diff
@@ -24,6 +27,13 @@ def __init__(self, name, params):
     def getOutputFiles(self):
         return self.specs['csvdiff']
 
+    # Check that override parameter lists are the same length
+#    def checkRunnable(self, options):
+#        if (len(self.specs['override_columns']) != len(self.specs['override_rel_err'])) or (len(self.specs['override_columns']) != len(self.specs['override_abs_zero'])) or (len(self.specs['override_rel_err']) != len(self.specs['override_abs_zero'])):
+#          self.addCaveats('Override inputs not the same length')
+#          self.setStatus(self.bucket_skip.status, self.bucket_fail)
+#          return False
```
Comment (brianmoose) [inline python/TestHarness/testers/CSVDiff.py]: If you add a `return FileTester.checkRunnable(self, options)` at the end (at the same level as the if) it seems to work for me.
If you uncomment this block, be sure to fix the indentation as well.

File: python/TestHarness/testers/CSVDiff.py [hunk 7]
```diff
@@ -24,6 +27,13 @@ def __init__(self, name, params):
     def getOutputFiles(self):
         return self.specs['csvdiff']
 
+    # Check that override parameter lists are the same length
+    def checkRunnable(self, options):
+        if (len(self.specs['override_columns']) != len(self.specs['override_rel_err'])) or (len(self.specs['override_columns']) != len(self.specs['override_abs_zero'])) or (len(self.specs['override_rel_err']) != len(self.specs['override_abs_zero'])):
+           self.addCaveats('Override inputs not the same length')
+           self.setStatus(self.bucket_skip.status, self.bucket_fail)
```
Comment (milljm) [inline python/TestHarness/testers/CSVDiff.py]: Do you mean to skip this test and supply a caveat as to the reason why? Or fail it?

Fail method:
```python
self.setStatus('Override inputs not the same length', self.bucket_fail)
```

Skip method:
```python
self.addCaveats('Override inputs not the same length')
self.setStatus(self.bucket_skip.status, self.bucket_skip)
```

**Review by brianmoose — DISMISSED** [review]

**Comment by acasagran** [comment]
@brianmoose Not sure why the dry_run python test is failing. Did I do something or is it just a glitch?

**Comment by brianmoose** [comment]
Yeah, we are talking about this.
Now that `checkRunnable` is correctly setting it to failed, dry run seems to handle this a bit differently than what we are used to. Most of the time `checkRunnable` marks them as skipped but yours marks it as failed.
You can either fix the dry run test or break out your `test_csvdiff` test to another test spec file and test that.
@milljm Any other suggestions?

**Comment by milljm** [comment]
I don't have a solution for checkRunnable to set failed statuses unfortunately.
These two options are clearly in conflict with each other. I am not sure which option we want to supersede the other.

I feel we should error, due to a parsing issue which should never enter the repository, even in the event dry-run normally returns a zero exit code. Which means I think we should adjust the unittest to pass for this one instance (use a different spec file for unittest).

**Review by brianmoose — APPROVED** [review]

**Pull request merged**